core/src/main/java/com/linecorp/armeria/server/http/auth/HttpAuthService.java

```diff
@@ -0,0 +1,63 @@
+package com.linecorp.armeria.server.http.auth;
+
+import java.util.AbstractMap.SimpleImmutableEntry;
+import java.util.Base64;
+import java.util.Base64.Decoder;
+import java.util.Map.Entry;
+import java.util.function.Function;
+import java.util.function.Predicate;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+import com.linecorp.armeria.common.http.DefaultHttpResponse;
+import com.linecorp.armeria.common.http.HttpHeaderNames;
+import com.linecorp.armeria.common.http.HttpRequest;
+import com.linecorp.armeria.common.http.HttpResponse;
+import com.linecorp.armeria.common.http.HttpStatus;
+import com.linecorp.armeria.server.DecoratingServiceFunction;
+
+final class HttpAuthService {
+    private static final Pattern AUTHORIZATION_HEADER_PATTERN = Pattern.compile(
+            "(?<mechanism>\\S+) (?<authorization>\\S+)");
+    private static final Decoder BASE_64_DECODER = Base64.getDecoder();
+
+    private static final Function<String, Entry<String, String>> basicAuthTokenExtractor = (String s) -> {
+        Matcher matcher = AUTHORIZATION_HEADER_PATTERN.matcher(s);
+        if (!matcher.matches()) {
+            throw new IllegalArgumentException("Invalid authorization header: " + s);
+        }
+
+        String mechanism = matcher.group("mechanism");
+
+        if (!"Basic".equals(mechanism)) {
+            throw new IllegalArgumentException("Invalid authorization mechanism: " + mechanism);
+        }
+
+        String authorization = new String(BASE_64_DECODER.decode(matcher.group("authorization")));
+        String[] tokens = authorization.split(":");
+        String user = tokens[0];
+        String password = tokens[1];
+
+        return new SimpleImmutableEntry<String, String>(user, password);
+    };
+
+    /**
+     * Returns DecoratingServiceFunction for basic authentication predicate.
+     * */
+    public static DecoratingServiceFunction<HttpRequest, HttpResponse> basic(
```

Comment:
`DecoratingServiceFunction` is an interface provided to users for convenience. We (Armeria devs) have to return a `Function` that decorates a service with a `DecoratingService`. See other decorating service implementations with static `newDecorator()`s.

The reason why we do this in spite of inconvenience on our side is because it gives a user better experience chaining decorators. For example, a user cannot do the following with `DecoratingServiceFunction`:

```java
SomeService s = myService.decorate(SomeService.newDecorator(...));
// Call the method declared in SomeService.
s.doSome();
```